Add new Nintendo Parental Controls integration

pantherale0 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new Nintendo Parental Controls integration, at the moment this is very basic, however future PRs will update the integration to support the platforms and features available in my custom integration (https://github.com/pantherale0/ha-nintendoparentalcontrols)

Library: https://github.com/pantherale0/pynintendoparental


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39118
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

pantherale0 committed:
Add nintendo parental integration

pantherale0 committed:
Add nintendo tests

pantherale0 committed:
Fix quality scale

pantherale0 committed:
Update pynintendoparental to version 0.6.7 and refactor imports in Nintendo parental controls integration

pantherale0 committed:
Update quality scale

pantherale0 committed:
Update homeassistant/components/nintendo_parental/entity.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/entity.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/sensor.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/sensor.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/sensor.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/config_flow.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 commented on homeassistant/components/nintendo_parental/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+                await self.auth.complete_login(
+                    self.auth, user_input[CONF_API_TOKEN], False
+                )
+            except (ValueError, InvalidSessionTokenException, HttpException):
+                errors["base"] = "invalid_auth"
+            else:
```
The library raises a value error when the provided URL is invalid, you can see that here https://github.com/pantherale0/pynintendoparental/blob/a7605f94f2ede7ece5bfaacc42eeb9cf7408c1f3/pynintendoparental/authenticator/__init__.py#L48

InvalidSessionTokenException is raised when the OAuth fails for some reason (invalid_grant etc) and Nintendo's auth service gives us a 400 error, this can happen if the user takes too long copying the URL into Home Assistant.

Thanks for the catch around general exceptions, I think I'll expand on the suggestion so the logger includes the trace otherwise errors will be difficult to fix.

pantherale0 committed:
Update homeassistant/components/nintendo_parental/config_flow.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/strings.json

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 commented on homeassistant/components/nintendo_parental/config_flow.py:
```diff
@@ -0,0 +1,86 @@
+                title=self.auth.account_id,
+                data={
+                    CONF_SESSION_TOKEN: self.auth.get_session_token,
+                    CONF_UPDATE_INTERVAL: user_input[CONF_UPDATE_INTERVAL],
```
Oh yes - had forgotten.. Will remove.

tr4nt0r commented:
Please run `python -m script.hassfest` to register the integration in Home Assistant

tr4nt0r commented:
And please also add the link to the library source (github) to the PR description

pantherale0 commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,82 @@
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """Return extra state attributes."""
+        if self.entity_description.state_attributes == "daily_summaries":
+            return {"daily": self._device.daily_summaries[0:5]}
+        return None
```
I will change to use a lambda for this because I don't want to store duplicate states (I've got another sensor to add as yet).

pantherale0 committed:
fixes

tr4nt0r commented on homeassistant/components/nintendo_parental/strings.json:
```diff
@@ -0,0 +1,30 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Click this link [Nintendo Login]({link}) to login to Nintendo. For the account you wish to link, right click on the red button \"Select this person\" and click \"Copy Link\".",
```
```suggestion
        "description": "To obtain your access token, click [Nintendo Login]({link}) to sign in to your Nintendo account. Then, for the account you want to link, right-click on the red **Select this person** button and choose **Copy Link Address**.",
```

tr4nt0r commented on homeassistant/components/nintendo_parental/strings.json:
```diff
@@ -0,0 +1,30 @@
+          "api_token": "Access token"
+        },
+        "data_description": {
+          "api_token": "The link copied from the Nintendo website"
```
```suggestion
          "api_token": "Paste the link copied from the Nintendo website"
```

tr4nt0r commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,60 @@
+UPDATE_INTERVAL = timedelta(seconds=60)
+
+
+class NintendoUpdateCoordinator(DataUpdateCoordinator):
```
```suggestion
class NintendoUpdateCoordinator(DataUpdateCoordinator[None]):
```

tr4nt0r commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,60 @@
+            authenticator, hass.config.time_zone, hass.config.language
+        )
+
+    async def _async_update_data(self):
```
```suggestion
    async def _async_update_data(self) -> None:
```

tr4nt0r commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,60 @@
+            raise ConfigEntryAuthFailed(err) from err
+        except TimeoutError as err:
+            raise UpdateFailed(err) from err
+        return False
```
```suggestion
```

pantherale0 committed:
Add extra sensor and change entity naming method

tr4nt0r commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,60 @@
+        """Update data from Nintendo's API."""
+        try:
+            with contextlib.suppress(InvalidSessionTokenException):
+                async with asyncio.timeout(self.update_interval.total_seconds() - 5):
```
if I'm not mistaken, the coordinator takes care of cancelling the task after 30s 
```suggestion
                async with asyncio.timeout(self.update_interval.total_seconds() - 5):
```

pantherale0 replied:
I can't see in the `DataUpdateCoordinator` class a specific timeout like this. There is some handling for timeout errors though so will remove my handling in this method.

tr4nt0r commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,60 @@
+                async with asyncio.timeout(self.update_interval.total_seconds() - 5):
+                    return await self.api.update()
+        except InvalidOAuthConfigurationException as err:
+            raise ConfigEntryAuthFailed(err) from err
```
This will trigger the auth flow, which should not be implemented in the initial PR. For now use `ConfigEntryError` instead. Also use exception translations. If the exception returns a user friendly error message you can include it with a `translation_placeholder`. Same applies for the `UpdateFailed` exceptions.
```suggestion
            raise ConfigEntryError(
                translation_domain=DOMAIN,
                translation_key="some_key",
            ) from err
```

pantherale0 committed:
update config flow tests

tr4nt0r commented on homeassistant/components/nintendo_parental/config_flow.py:
```diff
@@ -0,0 +1,58 @@
+
+    def __init__(self) -> None:
+        """Initialize a new config flow instance."""
+        self.auth: Authenticator = Authenticator.generate_login()
```
```suggestion
        self.auth = Authenticator.generate_login()
```

tr4nt0r commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,152 @@
+)
+
+
+async def test_full_flow(hass: HomeAssistant, mock_authenticator_client) -> None:
```
typing. Same for the other tests.
```suggestion
async def test_full_flow(hass: HomeAssistant, mock_authenticator_client: MagicMock) -> None:
```

tr4nt0r commented on tests/components/nintendo_parental/conftest.py:
Don't mock the internals of the library, instead mock only the front-facing methods you access in the config flow

pantherale0 replied:
If I don't mock the internals of the library, the tests won't pass because:

- The authentication methods require valid responses from either a user input (and what is mocked is not valid), or the OAuth service itself.
- The verifier code is generated on each new instance of a config flow
- Data is needed inside the library itself to allow other functions to work (which will be more important for future PRs)

tr4nt0r commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,152 @@
+    assert result["step_id"] == "user"
+    assert "link" in result["description_placeholders"]
+
+    mock_request_handler.return_value = {
+        "status": 400,
+        "text": "ERROR",
+        "json": {},
+        "headers": {"Content-Type": "application/json"},
+    }
```
just let the client raise instead of mocking the http request. Parametrize to also test the other exceptions.
```
    mock_authenticator_client.complete_login.side_effect = InvalidSessionTokenException
```

pantherale0 replied:
See my reply above, this won't work because of the validation done in the library.

pantherale0 replied:
I did just attempt this again, tests fail - 

```
2025-05-21 10:26:13.542 INFO     MainThread homeassistant.setup:setup.py:394 Setting up nintendo_parental
2025-05-21 10:26:13.545 DEBUG    MainThread pynintendoparental.authenticator:__init__.py:65 >> Init authenticator.
2025-05-21 10:26:13.545 DEBUG    MainThread pynintendoparental.authenticator:__init__.py:135 Refreshing access token.
2025-05-21 10:26:13.546 DEBUG    MainThread pynintendoparental:__init__.py:49 Received request to update data.
2025-05-21 10:26:13.547 DEBUG    MainThread pynintendoparental:api.py:46 Sending request to get_account_devices
2025-05-21 10:26:13.547 DEBUG    MainThread pynintendoparental:api.py:57 Built URL https://api-lp1.pctl.srv.nintendo.net/moon/v1/users/aabbccddee112233/devices?filter.device.activated.$eq=true
2025-05-21 10:26:13.573 ERROR    MainThread homeassistant.components.nintendo_parental.coordinator:update_coordinator.py:451 Unexpected error fetching nintendo_parental data
Traceback (most recent call last):
  File "/home/jordan/Documents/repos/home-assistant/homeassistant/helpers/update_coordinator.py", line 380, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/homeassistant/components/nintendo_parental/coordinator.py", line 58, in _async_update_data
    return await self.api.update()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/pynintendoparental/__init__.py", line 50, in update
    await self._get_devices()
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/pynintendoparental/__init__.py", line 33, in _get_devices
    response = await self._api.send_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
    )
    ^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/pynintendoparental/api.py", line 78, in send_request
    async with session.request(
               ~~~~~~~~~~~~~~~^
        method=e_point.get("method"),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        url=url,
        ^^^^^^^^
        json=body
        ^^^^^^^^^
    ) as response:
    ^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/client.py", line 1425, in __aenter__
    self._resp: _RetType = await self._coro
                           ^^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/client.py", line 703, in _request
    conn = await self._connector.connect(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        req, traces=traces, timeout=real_timeout
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/connector.py", line 548, in connect
    proto = await self._create_connection(req, traces, timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/connector.py", line 1056, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/connector.py", line 1375, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<7 lines>...
    )
    ^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohttp/connector.py", line 1115, in _wrap_create_connection
    sock = await aiohappyeyeballs.start_connection(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
    )
    ^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohappyeyeballs/impl.py", line 127, in start_connection
    raise first_exception
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohappyeyeballs/_staggered.py", line 132, in run_one_coro
    result = await coro_fn()
             ^^^^^^^^^^^^^^^
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/aiohappyeyeballs/impl.py", line 183, in _connect_sock
    sock = socket.socket(family=family, type=type_, proto=proto)
  File "/home/jordan/Documents/repos/home-assistant/venv/lib64/python3.13/site-packages/pytest_socket.py", line 89, in __new__
    raise SocketBlockedError()
pytest_socket.SocketBlockedError: A test tried to use socket.socket.
2025-05-21 10:26:13.578 DEBUG    MainThread homeassistant.components.nintendo_parental.coordinator:update_coordinator.py:460 Finished fetching nintendo_parental data in 0.032 seconds (success: False)
2025-05-21 10:26:13.578 DEBUG    MainThread homeassistant.config_entries:config_entries.py:802 Config entry 'aabbccddee112233' for nintendo_parental integration not ready yet: A test tried to use socket.socket.; Retrying in 5 seconds
2025-05-21 10:26:13.581 DEBUG    MainThread homeassistant.core:core.py:1546 Bus:Handling <Event component_loaded[L]: component=nintendo_parental>

 tests/components/nintendo_parental/test_config_flow.py::test_invalid_api_token ⨯                                                                                                                                                                                           75% ███████▌  
 tests/components/nintendo_parental/test_config_flow.py::test_general_error ✓                                                                                                                                                                                              100% ██████████

---------- coverage: platform linux, python 3.13.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
homeassistant/components/nintendo_parental/__init__.py         22      7    68%   31-34, 39-41, 48
homeassistant/components/nintendo_parental/config_flow.py      27      1    96%   38
homeassistant/components/nintendo_parental/const.py             3      0   100%
homeassistant/components/nintendo_parental/coordinator.py      27      1    96%   60
homeassistant/components/nintendo_parental/entity.py           16     16     0%   3-34
homeassistant/components/nintendo_parental/sensor.py           33     33     0%   3-102
-----------------------------------------------------------------------------------------
TOTAL                                                         128     58    55%

================================================================================================================================ short test summary info =================================================================================================================================
FAILED tests/components/nintendo_parental/test_config_flow.py::test_invalid_api_token - AssertionError: assert <FlowResultType.CREATE_ENTRY: 'create_entry'> is <FlowResultType.FORM: 'form'>
 +  where <FlowResultType.FORM: 'form'> = FlowResultType.FORM

Results (1.69s):
       3 passed
       1 failed
         - tests/components/nintendo_parental/test_config_flow.py:65 test_invalid_api_token
```

tr4nt0r commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,152 @@
+    assert result["errors"] == {"base": "invalid_auth"}
+
+    # clear the error and test retry
+    mock_request_handler.return_value = {
+        "status": 200,
+        "text": "OK",
+        "json": {
+            "session_token": "valid_token",
+            "expires_in": 3500,
+            "id": "aabbccddee112233",
+            "name": "Home Assistant Tester",
+        },
+        "headers": {"Content-Type": "application/json"},
+    }
```
and to clear the exception...
```
        mock_authenticator_client.complete_login.side_effect = None
```

pantherale0 replied:
As above

joostlek commented on homeassistant/components/nintendo_parental/__init__.py:
```diff
@@ -0,0 +1,48 @@
+    except InvalidOAuthConfigurationException as err:
+        raise ConfigEntryError(err) from err
+    entry.runtime_data = coordinator = NintendoUpdateCoordinator(
+        hass=hass, authenticator=nintendo_auth, config_entry=entry
```
I generally dislike named arguments in case like this

tr4nt0r commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,94 @@
+SENSOR_DESCRIPTIONS: tuple[NintendoParentalSensorEntityDescription, ...] = (
+    NintendoParentalSensorEntityDescription(
+        key="playing_time",
+        name="used screen time",
```
don't hardcode the entity name. I think it did not work as expected because the `translation_key` was missing
```suggestion
        name="used screen time",
```

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,40 @@
+from .coordinator import NintendoUpdateCoordinator
+
+
+class NintendoDevice(CoordinatorEntity):
```
```suggestion
class NintendoDevice(CoordinatorEntity[..]):
```
Add the coordinator type in the slot

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,40 @@
+    """Represent a Nintendo Switch."""
+
+    def __init__(
+        self, coordinator: NintendoUpdateCoordinator, device: Device, entity_id: str
```
Rather not use the variable name entity id as it already has multiple use cases

pantherale0 replied:
Thats not actually needed now as can get that from the entity description anyway so will remove.

pantherale0 commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,40 @@
+        await super().async_added_to_hass()
+        self._device.add_device_callback(self.async_write_ha_state)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self._device.name} {self.entity_description.name}"
```
Will Home Assistant prefix the name of the device in entity IDs etc automatically?

pantherale0 commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,94 @@
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda device: device.today_playing_time,
+        state_attributes_fn=lambda device: {"daily": device.daily_summaries[0:5]},
```
Daily summaries is a list of usage summary from Nintendo, the "[0:5]" selects the previous 5 days to limit the amount of data that is saved by the recorder.

It can be used to generate some advanced cards around what users are doing on the Switch etc. Its not actually converted to anything in the library however. In Home Assistant it produces a list of objects that look like this:

```
  - deviceId: 58fd8d614befcacf
    date: "2025-05-19"
    result: ACHIEVED
    playingTime: 3000
    exceededTime: 0
    disabledTime: 0
    miscTime: 300
    importantInfos: []
    notices: []
    observations: []
    playedApps:
      - applicationId: 010042D00D900000
        title: "LEGO® Star Wars™: The Skywalker Saga"
        imageUri:
          extraSmall: >-
            https://atum-img-lp1.cdn.nintendo.net/i/c/70d5df4d933a42029e4e752dea3070c7_3
          small: >-
            https://atum-img-lp1.cdn.nintendo.net/i/c/70d5df4d933a42029e4e752dea3070c7_128
          medium: >-
            https://atum-img-lp1.cdn.nintendo.net/i/c/70d5df4d933a42029e4e752dea3070c7_256
          large: >-
            https://atum-img-lp1.cdn.nintendo.net/i/c/70d5df4d933a42029e4e752dea3070c7_512
          extraLarge: >-
            https://atum-img-lp1.cdn.nintendo.net/i/c/70d5df4d933a42029e4e752dea3070c7_1024
        hasUgc: false
        shopUri: https://ec.nintendo.com/apps/010042d00d900000/GB?lang=en-GB
        firstPlayDate: "2025-03-01"
    anonymousPlayer: null
    devicePlayers:
      - playerId: 10004ebb_2321ef76_c43990a1_c1f5148a
        nickname: caiden
        imageUri: >-
          https://storage.googleapis.com/upload-4fcba637/device_players/16fbbef8-125a-42fb-b28b-59a31a807dcc.jpg?x=1740160724144
        playingTime: 2700
        playedApps:
          - applicationId: 010042D00D900000
            firstPlayDate: "2025-03-01"
            playingTime: 2700
    timeZoneUtcOffsetSeconds: 3600
    lastPlayedAt: 1747673812
    createdAt: 1747621132
    updatedAt: 1747717100
```

joostlek replied:
Let's not put this in the state machine, let's instead put this in a service call in a follow up

pantherale0 commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,94 @@
+    async_add_devices: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up the sensor platform."""
+    if entry.runtime_data.api.devices is not None:
```
I've had people raise issues in the past for the custom integration where they've configured this without actually configuring parental controls itself. Therefore devices could be none if they attempt to setup an account that is not managing any parental controls as a Switch will only appear in the Nintendo API if it is setup for parental controls.

joostlek replied:
But isn't that something we should guard for during the config flow?

pantherale0 commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,85 @@
+    NintendoParentalSensorEntityDescription(
+        key="playing_time",
+        native_unit_of_measurement="min",
+        state_attributes="daily_summaries",
```
Always the little things I forget..

Actually I think the class approach is cleaner so will implement

pantherale0 committed:
Update homeassistant/components/nintendo_parental/strings.json

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/strings.json

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/config_flow.py

Co-authored-by: Manu <4445816+tr4nt0r@users.noreply.github.com>

pantherale0 committed:
Update typing for tests

pantherale0 committed:
Suggestions from review

pantherale0 committed:
Fix exceptions structure

pantherale0 committed:
Assert timedelta in coordinator

pantherale0 commented:
Thanks both for your reviews and feedback. I think we might be close now? Only thing left that is outstanding is regarding patching internals.

pantherale0 committed:
Add missing MagicMock typing for mock_request_handler

pantherale0 commented:
This can be marked as a draft again - adding some further tests and changing how the python module handles API calls so internals don't need to be patched (https://github.com/pantherale0/pynintendoparental/pull/29)

joostlek commented on homeassistant/components/nintendo_parental/config_flow.py:
```diff
@@ -0,0 +1,58 @@
+            except (ValueError, InvalidSessionTokenException, HttpException):
+                errors["base"] = "invalid_auth"
+            except Exception as exc:
+                _LOGGER.exception("Unknown error during setup", exc_info=exc)
```
```suggestion
                _LOGGER.exception("Unknown error during setup")
```
I believe this already happens

joostlek commented on homeassistant/components/nintendo_parental/coordinator.py:
```diff
@@ -0,0 +1,62 @@
+            with contextlib.suppress(InvalidSessionTokenException):
+                if TYPE_CHECKING:
+                    assert isinstance(self.update_interval, timedelta)
+                async with asyncio.timeout(self.update_interval.total_seconds() - 5):
```
but why do we have this in the first place?

pantherale0 replied:
Theres a historical reason somewhere - that I cannot remember what it is...

The suppression? Or the timeout?

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __init__(self, coordinator: NintendoUpdateCoordinator, device: Device) -> None:
+        """Initialize."""
+        super().__init__(coordinator=coordinator)
```
```suggestion
        super().__init__(coordinator)
```

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,34 @@
+        super().__init__(coordinator=coordinator)
+        self._device = device
+        self._attr_unique_id = f"{device.device_id}_{self.entity_description.key}"
+        self._attr_has_entity_name = True
```
this one can be set outside of the constructor

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,34 @@
+            manufacturer="Nintendo",
+            name=device.name,
+            entry_type=DeviceEntryType.SERVICE,
+            sw_version=device.extra["device"]["firmwareVersion"]["displayedVersion"],
```
so its a service, but on the other hand we also show the device version?

pantherale0 replied:
Yes, a service because this doesn't technically interact with a device, but maybe I'm thinking to "technical" and actually it should be a device, because it does actually represent a device I suppose...

Shall I change to a device?

pantherale0 replied:
Changed to a device

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,34 @@
+            sw_version=device.extra["device"]["firmwareVersion"]["displayedVersion"],
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is loaded."""
+        await super().async_added_to_hass()
+        self._device.add_device_callback(self.async_write_ha_state)
```
We should also deregister this one

joostlek commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,102 @@
+    NintendoParentalSensorEntityDescription(
+        key=NintendoParentalSensor.TIME_REMAINING,
+        translation_key=NintendoParentalSensor.TIME_REMAINING,
+        native_unit_of_measurement="min",
```
please use UnitOfTime

joostlek commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,102 @@
+        description: NintendoParentalSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        self.entity_description = (
+            description  # this first so the NintendoDevice class has the entity key
+        )
+        super().__init__(coordinator=coordinator, device=device)
```
I would personally prefer to pass the key to the super constructor

pantherale0 committed:
Update homeassistant/components/nintendo_parental/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

pantherale0 committed:
Update homeassistant/components/nintendo_parental/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

pantherale0 committed:
improve error handling, update dependencies, and enhance entity management

pantherale0 committed:
Mark 'inject-websession' as done in quality_scale.yaml

pantherale0 committed:
Bump pynintendoparental for API v2

pantherale0 commented:
Converting this to draft as the API has changed following the release of the Switch 2 and currently awaiting on some testing (https://github.com/pantherale0/ha-nintendoparentalcontrols/issues/149)

pantherale0 committed:
Bump pynintendoparental

pantherale0 commented:
Ok, this is ready and working. A few changes on the Nintendo side that I've had to fix (first time in over 2 years the API has had any big changes and it appears it was only because of the release of the Switch 2).

My custom integration has broken for some reason, but this is tested and working fine now..

Just for clarity on the latest PR, Nintendo are now reporting usage in minutes rather than seconds so the `/60` was removed as its not required anymore.

Hopefully this can be merged in before the 2025.7.0 release so I can add additional platforms (no library changes required as the functionality already exists and is working there).

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,40 @@
+        await super().async_added_to_hass()
+        self._device.add_device_callback(self.async_write_ha_state)
+
+    async def async_removed_from_registry(self) -> None:
+        """When entity is removed."""
+        self._device.remove_device_callback(self.async_write_ha_state)
```
```suggestion
    async def async_removed_from_hass(self) -> None:
        """When entity is removed."""
        self._device.remove_device_callback(self.async_write_ha_state)
        await super().async_removed_from_hass()
```

joostlek commented on homeassistant/components/nintendo_parental/quality_scale.yaml:
```diff
@@ -0,0 +1,84 @@
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
+  entity-event-setup:
+    status: exempt
+    comment: |
+      No explicit event subscriptions.
```
well yes you do

pantherale0 replied:
Can I just mark this one done?

joostlek replied:
Not everything is event based, but you subscribe your entities to a device callback, which is an event

joostlek commented on homeassistant/components/nintendo_parental/quality_scale.yaml:
```diff
@@ -0,0 +1,84 @@
+  # Gold
+  devices: done
+  diagnostics: todo
+  discovery-update-info:
+    status: exempt
+    comment: |
+      No IP discovery.
+  discovery:
+    status: exempt
+    comment: |
+      No discovery.
```
why not?

pantherale0 replied:
This integration is cloud based, plus it requires further configuration between the Nintendo Switch and Parental Control app.

I didn't want this to appear as integration with the Nintendo Switch itself hence the exclusion of discovery for this one.

joostlek replied:
I think we can propose the user to add the intgration when we find a nintendo switch on the network? It's kinda similar to Playstation network I guess

joostlek commented on homeassistant/components/nintendo_parental/sensor.py:
```diff
@@ -0,0 +1,91 @@
+
+
+SENSOR_DESCRIPTIONS: tuple[NintendoParentalSensorEntityDescription, ...] = (
+    NintendoParentalSensorEntityDescription(
+        key=NintendoParentalSensor.PLAYING_TIME,
+        translation_key=NintendoParentalSensor.PLAYING_TIME,
+        native_unit_of_measurement=UnitOfTime.MINUTES,
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda device: device.today_playing_time,
+    ),
+    NintendoParentalSensorEntityDescription(
+        key=NintendoParentalSensor.TIME_REMAINING,
+        translation_key=NintendoParentalSensor.TIME_REMAINING,
+        native_unit_of_measurement=UnitOfTime.MINUTES,
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda device: device.today_time_remaining,
+    ),
```
we could add a suggested unit of measurement if that fits

pantherale0 replied:
These are always in minutes when returned from the API itself. Not sure if there is a benefit to adding a suggested unit of measurement?

joostlek replied:
I am wondering if most users consume this in minutes or hours

joostlek commented on homeassistant/components/nintendo_parental/strings.json:
```diff
@@ -0,0 +1,38 @@
+          "api_token": "Access token"
+        },
+        "data_description": {
+          "api_token": "Paste the link copied from the Nintendo website"
```
This is now explaining an action while it should be explaining the field

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,119 @@
+def mock_request_handler() -> Generator[MagicMock]:
+    """Mock the _request_handler function."""
+    with patch(
+        "pynintendoparental.authenticator.Authenticator._request_handler",
```
1. we should patch this where we use it
2. Why to we patch itnernals?

pantherale0 replied:
1. good question
2. good question

Wrote these tests with duck tape over my eyes ;) but I've changed a fair bit in the Python module itself.... All of my py modules hid all of the HTTP methods behind something really generic, at the time it was a case of "adding a new api endpoint will be easy this way"... a couple of years later its proved to be a nightmare when debugging.

I will rewrite these.

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,119 @@
+    """Mock a Nintendo authenticator."""
+    with (
+        patch(
+            "pynintendoparental.authenticator.Authenticator",
+            autospec=True,
+        ) as mock_nintendo_auth,
+        patch(
+            "pynintendoparental.authenticator._parse_response_token",
```
idem

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,119 @@
+        yield mock_nintendo_auth
+
+
+@pytest.fixture
+def mock_nintendo_device() -> Generator[MagicMock]:
+    """Mock a single device."""
+    with patch(
+        "pynintendoparental.device.Device", autospec=True, return_value=True
+    ) as mock_nintendo_device:
+        mock_nintendo_device.device_id.return_value = "testdevid"
+        mock_nintendo_device.name.return_value = "Home Assistant Test"
+        mock_nintendo_device.extra.return_value = {
+            "device": {"firmwareVersion": {"displayedVersion": "99.99.99"}}
+        }
+        mock_nintendo_device.limit_time.return_value = 120
+        mock_nintendo_device.today_playing_time.return_value = 110
+        yield mock_nintendo_device
+
+
+@pytest.fixture
+def mock_nintendo_devices(mock_nintendo_device: MagicMock) -> dict[str, Device]:
+    """Mock a collection of devices."""
+    return {"testdevid": mock_nintendo_device}
+
+
+@pytest.fixture
+def mock_nintendo_client(
+    mock_nintendo_client_init: MagicMock,
+    mock_nintendo_devices: MagicMock,
+    mock_authenticator_client: MagicMock,
+) -> Generator[AsyncMock]:
+    """Mock a Nintendo client."""
+    client = mock_nintendo_client_init.return_value
+    client.update.return_value = True
+    client.devices.return_value = mock_nintendo_devices.return_value
+    return client
```
If you want to create a mock, instead of using patch and patch the library, create a mock and autospec it. Check `niko_home_control` for examples

joostlek commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+
+from tests.common import MockConfigEntry
+
+mock_config_entry = MockConfigEntry(
+    domain=DOMAIN,
+    data={CONF_SESSION_TOKEN: "valid_token"},
+    unique_id="aabbccddee112233",
+)
```
you can make this a fixture as you will use it in more places

joostlek commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+) -> None:
+    """Test a full and successful config flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

pantherale0 committed:
Update homeassistant/components/nintendo_parental/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

emontnemery commented:
@pantherale0 feel free to message me on Discord if you're stuck with the tests and need some help with it

pantherale0 committed:
Refactor tests

pantherale0 committed:
lint

pantherale0 committed:
Use async_removed_from_registry

pantherale0 committed:
Use async_will_remove_from_hass

pantherale0 commented:
I don't think the CI error is from my integration, but I'll double check.

pantherale0 committed:
Merge branch 'dev' into integration/nintento_parental

pantherale0 committed:
Refactor Nintendo parental controls config flow and improve error handling

joostlek commented on homeassistant/components/nintendo_parental/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+            )
+
+        if user_input is not None:
+            try:
+                await self.auth.complete_login(
+                    self.auth, user_input[CONF_API_TOKEN], False
+                )
+                if TYPE_CHECKING:
+                    assert self.auth.account_id
+                await self.async_set_unique_id(self.auth.account_id)
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=self.auth.account_id,
+                    data={
+                        CONF_SESSION_TOKEN: self.auth.get_session_token,
+                    },
+                )
+            except (ValueError, InvalidSessionTokenException, HttpException):
+                errors["base"] = "invalid_auth"
```
Only have things in the try block that can raise

joostlek commented on homeassistant/components/nintendo_parental/entity.py:
```diff
@@ -0,0 +1,40 @@
+
+    async def async_will_remove_from_hass(self) -> None:
+        """When will be removed from HASS."""
+        self._device.remove_device_callback(self.async_write_ha_state)
```
```suggestion
        self._device.remove_device_callback(self.async_write_ha_state)
        await super().async_will_remove_from_hass()
```

joostlek commented on homeassistant/components/nintendo_parental/strings.json:
```diff
@@ -0,0 +1,38 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "To obtain your access token, click [Nintendo Login]({link}) to sign in to your Nintendo account. Then, for the account you want to link, right-click on the red **Select this person** button and choose **Copy Link Address**.",
```
Not everyone sets this up on desktop tho

pantherale0 replied:
I'm not sure how best to address this - its stated in the docs that you need to remove the Nintendo Switch Parental Controls app first (and the reason why) https://github.com/home-assistant/home-assistant.io/blob/fdd531e60564e5e34ceba216955a40e6931d9ee3/source/_integrations/nintendo_parental.markdown

"open the context menu" sounds even more confusing, but you need that link to be able to proceed.

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,89 @@
+def mock_config_entry() -> MockConfigEntry:
+    """Return a mock config entry."""
+    return MockConfigEntry(
+        domain="nintendo_parental",
```
```suggestion
        domain=DOMAIN,
```

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,89 @@
+    """Return a mock config entry."""
+    return MockConfigEntry(
+        domain="nintendo_parental",
+        data={"session_token": "valid_token"},
```
can we use constants?

joostlek commented on tests/components/nintendo_parental/conftest.py:
```diff
@@ -0,0 +1,89 @@
+    return MockConfigEntry(
+        domain="nintendo_parental",
+        data={"session_token": "valid_token"},
+        unique_id="aabbccddee112233",
```
can we pick a more realistic unique id?

joostlek commented on tests/components/nintendo_parental/test_config_flow.py:
```diff
@@ -0,0 +1,86 @@
+        result["flow_id"], user_input={CONF_API_TOKEN: "invalid_token"}
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
+    assert result["errors"] == {"base": "invalid_auth"}
```
We should finish this flow in CREATE_ENTRY, not just FORM to test that the flow is able to recover

pantherale0 committed:
Update homeassistant/components/nintendo_parental/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

pantherale0 committed:
Update tests/components/nintendo_parental/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

pantherale0 committed:
updates from review

quality scale
consts in tests
move non-exception code outside of try statement in config flow

pantherale0 committed:
ruff
